content/blog/posts/2022-08-04-add-desktop-notifications.md

```diff
@@ -0,0 +1,241 @@
+
+```js
+async _isShowNotificationPermitted() {
+  if ('Notification' in window) {
+    let { permission } = Notification;
+    if (permission !== 'granted') {
+      permission =  await Notification.requestPermission();
+    }
+    if (permission !== 'denied') {
+      return true;
+    }
+  }
+
+  return false;
+}
+```
+
+Let's break down this method a bit. First we're checking for support in the browser, in the off chance your user doesn't have a supported browser, otherwise why have those needless errors? The default value from `Notification.permission` is just that, `default`, which we treat essentially the same as `denied`. In that case we trigger `requestPermission()` in the API to request access from the user. If the user approves, we can move on to showing them notifications until this permission is revoked. Let's look then at a method for showing one.
+
+```js
+async _showNotification() {
+  const permission = await this._isShowNotificationPermitted();
+
+  if (permission !== 'denied') {
+    new Notification('Be notified of new posts', {
+      body: 'Let us notifiy you of new posts on our site!',
+    });
+  }
+}
+```
+
+This one is a bit more straightforward. We check permission, as noted in the previous code example, and as long as we're not denied, a notification can be shown. There's really only one required argument, a title. Here we are adding a basic option for body text as well. This will just show a notification and the user can dismiss it or select to go to browser settings, if they'd like to change the permissions. Let's make sure our instance fires off when loaded before we move on.
+
+```js
+class NotificationService {
+  constructor() {
+    this._showNotification();
+  }
+
+// ...the rest of the class
+```
+
+![Basic desktop notification](/img/blog/add-desktop-notifications/basic-notification.png)
+
+This gets us a basic notification, but without any interaction from the user. To do that, we have to add actions to the options hash that you can watch and respond too. The spec requires that we only can do that if we're triggering the notification via a service worker and responding to events there. Let's create a basic service worker file and add a listener. Create a file called `service-worker.js` and add the following code.
+
+```js
+self.addEventListener('notificationclick', function(event) {
+  const clickedNotification = event.notification;
+  clickedNotification.close();
+
+  // Do something as the result of the notification click
+  const promiseChain = doSomething();
+  event.waitUntil(promiseChain);
+});
+```
+
+In this we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.
```

Comment:
```suggestion
In this snippet we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.
```

Reply:
stylistic choice but naming what we are referring comes off more direct and clear